test/test_ratelimit.py

```diff
@@ -0,0 +1,155 @@
+def time():
+    with patch("time.time") as time_mock:
+        time = ControllableTime()
+        time_mock.side_effect = time
+        yield time
+
+
+def test_resource_ratelimit_bucket(time):
+    time.set_to(1000)
+    bucket = ResourceRateLimitBucket("abc123")
+    bucket.update(make_ratelimit_headers(reset=1001, reset_after=1))
+
+    assert bucket.bucket_id == "abc123"
+    assert bucket.reset == 1001
+    assert bucket.remaining == 1
+
+    bucket.take()  # doesn't raise
+    assert bucket.remaining == 0
+
+    with pytest.raises(RateLimitException) as exc_info:
+        bucket.take()
+
+    e = exc_info.value
+    assert e.reset == 1001 and e.is_global == False
+
+
+def test_global_ratelimit_bucket(time):
+    time.set_to(100)
+    limit = GlobalRateLimitBucket()
+
+    assert not limit.is_ratelimited
+    limit.take()  # doesn't raise
+
+    limit.update(make_global_ratelimit_headers(100))
+
+    assert limit.is_ratelimited
+
+    with pytest.raises(RateLimitException) as exc_info:
+        limit.take()
+
+    e = exc_info.value
+    assert e.reset == 101 and e.is_global == True
+
+
+def test_ratelimit_passes_first_request():
+    limiter = RateLimiter()
+    limiter.on_request(default_request)  # doesn't raise
+
+
+def test_ratelimit_passes_good_response(time):
+    response = make_response(200, make_ratelimit_headers())
+    limiter = RateLimiter()
+    limiter.on_response(response)  # doesn't raise
+
+
+@pytest.mark.parametrize(
+    "start, reset, is_global, response",
+    [
+        (0, 1, True, make_response(429, make_global_ratelimit_headers(100))),
+        (
+            1000,
+            1001,
+            False,
+            make_response(429, make_ratelimit_headers("abc123", 10, 0, 1001, 1)),
+        ),
+    ],
+)
+def test_ratelimit_raises_on_limit_exhausted_response(
+    time, start, reset, is_global, response
+):
+    time.set_to(start)
+    limiter = RateLimiter()
+    with pytest.raises(RateLimitException) as exc_info:
+        limiter.on_response(response)
+
+    e = exc_info.value
+    assert e.reset == reset and e.is_global == is_global
+
+
+def test_ratelimit_raises_on_request_exhausted_resource(time):
+    time.set_to(1000)
+    limiter = RateLimiter()
+    bucket = limiter.resource_buckets[
+        (default_request.method, default_request.url)
+    ] = ResourceRateLimitBucket("abc123")
+    bucket.update(make_ratelimit_headers("abc123", 10, 0, 1002, 2))
+
+    with pytest.raises(RateLimitException) as exc_info:
+        bucket.take()
+
+    assert exc_info.value.reset == 1002
+
+
+@pytest.mark.parametrize(
+    "url, group",
+    [
+        ("channels/2909267986263572999", "channels/2909267986263572999"),
+        ("guilds/197038439483310086", "guilds/197038439483310086"),
+        ("webhooks/223704706495545344", "webhooks/223704706495545344"),
+        ("/channels/2909267986263572999/", "channels/2909267986263572999"),
+        ("/guilds/197038439483310086/members/{user.id}", "guilds/197038439483310086"),
```

Comment:
Am I understanding that any url starting with `/guilds/197038439483310086` would be mapped to the same bucket?